core/src/main/java/brooklyn/management/internal/BasicExternalConfigSupplierRegistry.java

```diff
@@ -0,0 +1,108 @@
+        if (providersByName.containsKey(name))
+            throw new IllegalArgumentException("Provider already registered with name '" + name + "'");
+        providersByName.put(name, provider);
+    }
+
+    synchronized public void removeProvider(String name) {
+        providersByName.remove(name);
+    }
+
+    /**
+     * Searches the named {@link ExternalConfigSupplier} for the config value associated with the specified key.
+     * Quietly returns <code>null</code> if no config exists for the specified key.
+     * Throws {@link IllegalArgumentException} if no {@link ExternalConfigSupplier} exists for the passed name.
+     */
+    synchronized public String getConfig(String providerName, String key) {
+        ExternalConfigSupplier provider = providersByName.get(providerName);
+        if (provider == null)
+            throw new IllegalArgumentException("No provider found with name '" + providerName + "'");
+        return provider.get(key);
+    }
+
+    @SuppressWarnings("unchecked")
+    synchronized private void updateFromBrooklynProperties(ManagementContext mgmt) {
+        // form is:
+        //     brooklyn.external.<name> : fully.qualified.ClassName
+        //     brooklyn.external.<name>.<key> : <value>
+        //     brooklyn.external.<name>.<key> : <value>
+        //     brooklyn.external.<name>.<key> : <value>
+
+        String EXTERNAL_PROVIDER_PREFIX = "brooklyn.external.";
+        ConfigMap externalProviderProperties = mgmt.getConfig().submap(ConfigPredicates.startingWith(EXTERNAL_PROVIDER_PREFIX));
+        ClassLoader classloader = mgmt.getCatalogClassLoader();
+
+        for (String key : externalProviderProperties.asMapWithStringKeys().keySet()) {
+            String strippedKey = key.substring(EXTERNAL_PROVIDER_PREFIX.length());
+            if (strippedKey.contains("."))
+                continue;
+
+            String name = strippedKey;
+            String providerClassname = (String) externalProviderProperties.asMapWithStringKeys().get(key);
+            Map<String, Object> config = ConfigUtils.filterForPrefixAndStrip(externalProviderProperties.asMapWithStringKeys(), key+".");
+
+            try {
+                Class<? extends ExternalConfigSupplier> providerClass = (Class<? extends ExternalConfigSupplier>) classloader.loadClass(providerClassname);
+                Constructor<? extends ExternalConfigSupplier> constructor = providerClass.getConstructor(
+                    ManagementContext.class, String.class, Map.class);
+                ExternalConfigSupplier configSupplier = constructor.newInstance(this, name, config);
+                addProvider(name, configSupplier);
+                LOG.info("Added external config supplier named '" + name + "': " + configSupplier);
+
+            } catch (Exception e) {
+                LOG.error("Failed to instantiate external config supplier named '" + name + "': " + e, e);
```

Comment:
Should we just log and continue? Or should we propagate? Given these values came from brooklyn.properties (rather than persisted state etc), then it feels reasonable to fail startup of the management context.

Thoughts?